[[ Bug 11039 ]] throw error when changing executing behavior

montegoulding opened the pull request:
Now that I noticed there was a partial fix to the issue on develop I have cherry-picked and then added a fix for when setting to empty.

livecodeali committed:
[[ Bug 11039 ]] Throw error when changing behavior in behavior
This patch prevents behavior scripts from changing their child's
behavior.

(cherry picked from commit 24e482a1c75add479362a3501a1df4180d85e1a9)

livecodeali committed:
[[ Bug 11039 ]] Add more detail to release note

(cherry picked from commit 762f616937875033ef0fb2753ceca5b653f099f8)

livecodeali committed:
[[ Bug 11039 ]] Improve SetParentScript implementation
Use an autopointer to cleanup parsed chunk, to enable throw/return
when object fails various conditions. Also return doing nothing if
the parent script object is not being changed, even if it is
currently executing code.

(cherry picked from commit 044d78293d30a3c8ca74922ea0f2f2870d1aceba)

livecodeali committed:
[[ Tests ]] Add behavior tests
- Add tests for remaining execution error
- Add tests for general behavior functionality

(cherry picked from commit 438af978b713d26ea3eb949faf6af6090282151f)

livecodeali commented on tests/lcs/core/engine/behavior.livecodescript:
```diff
@@ -148,4 +148,26 @@ on TestBehaviorHandlerResolveOrder
+   set the script of button "child" to "on Foo;Bar;end Foo"
+   create button "parent"
+   set the script of button "parent" to "on Bar;Baz;end Bar"
+   set the behavior of button "child" to the long id of button "parent"
```
Is this line meant to be here? The first test seems to be saying the behavior starts off empty

montegoulding replied:
Ah good catch!

montegoulding committed:
[[ Bug 11039 ]] Ensure set executing behavior to empty throws

This patch ensures that setting the behavior of an object to empty while
the behavior script is executing throws an error. A previous patch ensured
that setting to a sifferent object throws. If no error is thrown and there
is an error thrown in the message path then the invalid behavior is still
in the execution contexts causing a crash.

livecodeali commented on tests/lcs/core/engine/behavior.livecodescript:
```diff
@@ -148,4 +148,26 @@ on TestBehaviorHandlerResolveOrder
+   set the script of stack "test" to "on Bar;set the behavior of button " & quote & "child" & quote & " to the long id of button " & quote & "parent"& quote &";end Bar"
+   TestAssertDoesNotThrow "can set behavior while executing if it is empty", "Foo", \
+      the long id of button "child"
+ 
```
Now it looks like there's a missing 
```
set the behavior of button "child" to the long id of button "parent"
```
here

montegoulding replied:
The test does it

livecodeali replied:
doh!

livecodeali commented:
@livecode-vulcan review ok 15a880e

livecode-vulcan commented:
💙 review by @livecodeali ok 15a880ee02692fb9f90bd3ff1260911cf950ea66

livecode-vulcan commented:
😎 test success 15a880ee02692fb9f90bd3ff1260911cf950ea66

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1029)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1045)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1019)
* try-community-universal-ios-iphoneos11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.1/builds/92)
* try-community-universal-ios-iphonesimulator11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.1/builds/92)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1024)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1046)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1029)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1045)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1028)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1038)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1061)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1024)
